Adding orderToIndex routine for rectangular domains

CaptainSharf opened the pull request:
As part of this PR
- Adding the routine `orderToIndex` for rectangular domains which returns an N-dimensional index for an `order`
- Adding tests to verify the functionality of this routine

CaptainSharf commented:
Hi @bradcray , @ben-albrecht ,
Could you please verify if the routine `orderToIndex` implemented is valid and meets the necessary requirements for dense rectangular domains?

CaptainSharf commented:
This issue has been fixed as part of the PR #16514

bradcray commented:
I'm interested in taking a look at this, but probably won't have time to do so until after next week's code freeze (even if someone else were to have time to review it today, I doubt we would merge it today because the feature freeze for this release was this past Monday).

CaptainSharf commented:
Hi @bradcray , @ben-albrecht could you please take a look at this PR if you're free?

bradcray commented:
Still working on getting tomorrow's release out the door, sorry for the continuing delay...  It's the busy season.

ben-albrecht reviewed (changes requested):
@CaptainSharf - This looks pretty good. See my feedback inline, including some trivial stylistic nitpicks to be more consistent with the rest of `ChapelArray`.

ben-albrecht commented on modules/internal/ChapelArray.chpl:
```diff
@@ -1887,19 +1887,22 @@ module ChapelArray {
+          if stride < 0 then 
+            currInd += hi;
+          else
+            currInd += lo;
```
I think the previous else block remained in after you merged the suggestions (I think I should have done a multi-line suggestion)

CaptainSharf replied:
Oh, no problem, I'll fix it

CaptainSharf commented on modules/internal/ChapelArray.chpl:
```diff
@@ -1885,6 +1885,28 @@ module ChapelArray {
     pragma "no doc"
     proc indexOrder(i) return _value.dsiIndexOrder(_makeIndexTuple(rank, i));
 
+    pragma "no doc"
```
@ben-albrecht , added a doc string for this method. Please let me know if I need to make changes to it

ben-albrecht commented on modules/internal/ChapelArray.chpl:
```diff
@@ -1885,6 +1885,50 @@ module ChapelArray {
     proc indexOrder(i) return _value.dsiIndexOrder(_makeIndexTuple(rank, i));
 
+    /*
+       Returns domain index for a given order in the domain
```
Let's reword this line and add some simple examples. Here's a suggested text:

```
Returns the `ith` index in the domain. For example, the `2nd` index of ``{2..10 by 2}`` would be ``4``.

The order of multidimensional domains follows their 1-dimensional iterator.
For example, the `4th` index of ``{1..3, 1..2}`` would be ``(2,1)``.
```

vasslitvinov replied:
"1-dimensional iterator for multidimensional domains" does not ring a bell to me.
Did you mean "serial iterator" ?

Minor: let us go with singular: `The order of [for?] a multidimensional domain follows` ......

ben-albrecht replied:
>Did you mean "serial iterator"

Yes, thanks for catching that.

@CaptainSharf - combining @vasslitvinov's feedback into my previous suggestion:

```
Returns the `ith` index in the domain. For example, the `2nd` index of ``{2..10 by 2}`` would be ``4``.

The order of a multidimensional domain follows the serial iterator.
For example, the `4th` index of ``{1..3, 1..2}`` would be ``(2,1)``.
```

CaptainSharf replied:
@ben-albrecht , wouldn't the `orderToIndex` for the second domain be `(2,2)` , as we're using 0 based indexing here?

CaptainSharf replied:
I also wanted to verify the behavior of this method when the rank of the domain is 1. In the first example, the method returns a 1-tuple i.e `(6)` .  Would this be acceptable?

ben-albrecht replied:
> @ben-albrecht , wouldn't the `orderToIndex` for the second domain be `(2,2)` , as we're using 0 based indexing here?

Oops, yes you are right. Good catch! I suppose the test would catch this too :)

ben-albrecht replied:
> In the first example, the method returns a 1-tuple i.e (6) . Would this be acceptable?

Good question -- I was intuitively expecting an integer when rank is 1. Is there a rationale for making it a 1-tuple? Maybe we ought to bring this up in https://github.com/chapel-lang/chapel/issues/14885.

ben-albrecht replied:
Sounds like there is consensus to always returning the `idxType` in https://github.com/chapel-lang/chapel/issues/14885. This means rank 1 domains just return the `idxType` instead of a 1-tuple, e.g. `6`, not `(6,)`.

ben-albrecht commented on test/domains/userAPI/test2DdomainOrderToIndex.chpl:
```diff
@@ -0,0 +1,29 @@
+}
+
+{
+    var dom = {1..3,1..3,1..5};
```
2-space indentation is the norm in Chapel development

ben-albrecht commented on test/domains/userAPI/test2DdomainOrderToIndex.chpl:
```diff
@@ -0,0 +1,29 @@
+{
+    var dom = {1..3,1..3};
+    testDom(dom);
+}
```
It would be a good idea to add the specific examples from the documentation into this test file to ensure future changes don't break the documentation examples. We can annotate these test cases with a reminder of why they are here:

```chpl
{
  // Example mentioned in the documentation
  var dom = {2..10 by 2};
  const idx = 4;
  const i = 2;
  if idx != dom.orderToIndex(i) {
    writeln('wrong index for ', dom, ':', idx, ' != dom.orderToIndex(', i,')');
    writeln('  got ', dom.orderToIndex(i), ' instead');
  }
}

{
  // <other example here>
}
```

ben-albrecht reviewed (changes requested):
@CaptainSharf - we are catching up on some reviews. Thanks for your patience on this PR.

This feature looks good to me overall barring some more stylistic changes and documentation improvements. See comments inline. Thanks!

ben-albrecht commented:
Linking this PR to the relevant issue: https://github.com/chapel-lang/chapel/issues/14885

CaptainSharf commented:
Thanks @ben-albrecht for the suggested changes! I'll make these improvements today!

mppf commented:
@CaptainSharf - I'm not sure what `git` operations you did but it seems you've added the commits from `master` to your branch with additional sign-off information in the commits (signed off by you). That wouldn't be OK to merge but that's besides the point because it's almost certainly not what you intended. I noticed the issue because for whatever reason GitHub emailed me for each of these commits mentioning me.

I recommend you `rebase` to get changes from master on to your current branch (but other people on the team prefer to `merge`). See https://github.com/chapel-lang/chapel/blob/master/doc/rst/developer/bestPractices/git.rst and the section starting with "There are two different strategies available to bringing changes from Chapel master into your development branch".

In any case at this point your branch might be too messed up to work with. If that's the case I recommend you:
 1. copy your changes to something outside of your tree (since you added a new routine you could just copy the new code somewhere; alternatively you could copy all of ChapelArray. I do not remember which files you had modified)
 2. check out a new version of `upstream` - `git fetch upstream && git checkout upstream/master` -- this will create an unnamed branch
     * if you don't have `upstream` set up you can set it up first with `git remote add upstream https://github.com/chapel-lang/chapel.git`
 3. rename your messed up branch to something else `git branch -m orderToIndex orderToIndex-bad`
 4. rename the unnnamed branch `git checkout -b orderToIndex`
 5. apply the changes you saved and commit
 6. push your fixed branch to github with `git push -f origin orderToIndex`
 
If you already have a branch where you've fixed things you can check that out and then go right to the renaming step (3).

CaptainSharf commented:
Hi, @mppf , Thanks for the pointers! I feel that this PR has far too many conflicts to work with now. @ben-albrecht , I've created a new PR #16959 which has my changes. I've added the procedure `orderToIndex` as well as tests for the same in it. I've also incorporated the review comments from this PR.

ben-albrecht commented:
Closed since #16959 is opened.
